### PR TITLE
ActiveAE: fix destructor

### PR DESCRIPTION
**Pull request by repojohnray**
## Description
This commit fixes some issues detected by the sanitizer which generate some memory leaks.

## Motivation and context
The goal is to clean up some issues detected by the sanitizer.

## How has this been tested?
-DECM_ENABLE_SANITIZERS='leak' exists, but it does not pass the proper option to the linker.

You could add the options below for the tests:
cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_FLAGS_DEBUG="-ggdb -O2 -fno-omit-frame-pointer -fsanitize=leak" -DCMAKE_EXE_LINKER_FLAGS="-fsanitize=leak" -DCMAKE_SHARED_LINKER_FLAGS="-fsanitize=leak"

The python library interferes with the sanitizer, "export PYTHONMALLOC=malloc" is not sufficient to fix this issue. Removing "runner.py" could be sufficient to fix this problem, otherwise disabling CPythonInvoker() is required.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp [hunk 1]
```diff
@@ -282,6 +282,21 @@ CActiveAE::~CActiveAE()
...
+
+  if (m_encoderBuffers)
+    delete m_encoderBuffers;
+
+  if (m_vizBuffers)
+    delete m_vizBuffers;
+
+  if (m_vizBuffersInput)
+    delete m_vizBuffersInput;
+
```
Comment (ksooo) [inline xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp:299]: No need for the ifs here. It is safe to call `delete` on a `nullptr`.

**Comment by lrusak** [comment]
I did a quick commit to use smart pointers for the buffer pools. It turned out pretty simply actually (only build tested not runtime tested). https://github.com/lrusak/xbmc/commit/d5f6e35c2b6668f4eb76b9b37cc63d4b133f7e8e

**Comment by repojohnray** [comment]
You are right about the ifs. I let you fix this issue with smart pointers.

**Comment by repojohnray** [comment]
@lrusak, your commit is fine and should replace mine.

**Comment by lrusak** [comment]
> @lrusak, your commit is fine and should replace mine.

Do you want to test it the same way to make sure the mem leaks are handled?

If you feel comfortable you can cherry-pick the commit and push that for this PR.

It needs proper runtime testing though as I haven't had a chance to test it yet.

**Comment by repojohnray** [comment]
The memory leaks are gone, anyway you are right it needs some runtime testing.
You should push your PR; The current PR is not needed anymore.

**Comment by neo1973** [comment]
> -DECM_ENABLE_SANITIZERS='leak' exists, but it does not pass the proper option to the linker.

What options are missing?

CMake:
```
CC=clang CXX=clang++ cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_INSTALL_LIBDIR=/usr/lib -DENABLE_AIRTUNES=ON -DENABLE_ALSA=ON -DENABLE_AVAHI=ON -DENABLE_BLURAY=ON -DENABLE_CEC=ON -DENABLE_DBUS=ON -DENABLE_DVDCSS=ON -DENABLE_EGL=ON -DENABLE_EVENTCLIENTS=ON -DENABLE_MICROHTTPD=ON -DENABLE_MYSQLCLIENT=ON -DENABLE_NFS=ON -DENABLE_OPTICAL=ON -DENABLE_PULSEAUDIO=ON -DENABLE_SMBCLIENT=ON -DENABLE_UDEV=ON -DENABLE_UPNP=ON -DENABLE_VAAPI=ON -DENABLE_XSLT=ON -DENABLE_INTERNAL_FFMPEG=ON -DENABLE_MOLD=ON -DCORE_PLATFORM_NAME=wayland -DAPP_RENDER_SYSTEM=gl -DECM_ENABLE_SANITIZERS='leak' -G Ninja ../xbmc/
```

Building the object file:
```
/usr/bin/clang++
[...]
-fsanitize=leak -gsplit-dwarf -Wall -Wdouble-promotion -Wmissing-field-initializers -Wsign-compare -Wextra -Wno-unused-parameter -Wno-bad-function-cast -Wno-deprecated -Wnon-virtual-dtor -g
[...]
-Werror=double-promotion -Werror=missing-field-initializers -Werror=sign-compare -msse -msse2 -std=c++17 -MD -MT build/cores/audioengine/CMakeFiles/audioengine.dir/Engines/ActiveAE/ActiveAE.cpp.o -MF build/cores/audioengine/CMakeFiles/audioengine.dir/Engines/ActiveAE/ActiveAE.cpp.o.d -o build/cores/audioengine/CMakeFiles/audioengine.dir/Engines/ActiveAE/ActiveAE.cpp.o -c /home/mark/Coding/Repos/kodi-git/xbmc/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp
```
Linking the executable:
```
/usr/bin/clang++ -fsanitize=leak -gsplit-dwarf -Wall -Wdouble-promotion -Wmissing-field-initializers -Wsign-compare -Wextra -Wno-unused-parameter -Wno-bad-function-cast -Wno-deprecated -Wnon-virtual-dtor -g -D_DEBUG -fuse-ld=mold -fuse-ld=mold  -Wl,--gdb-index   -Wl,--export-dynamic -rdynamic CMakeFiles/kodi.dir/xbmc/platform/posix/main.cpp.o -o kodi-wayland  -Wl,--whole-archive  [...]
```
Both contain `-fsanitize=leak`.

**Comment by repojohnray** [comment]
@neo1973,  do you get the proper executable with debuginfo enabled once the compilation/linkage is over?

Another issue, if you want a fully sanitized executable, you could use the following options: "-fsanitize=bounds -fsanitize=address -fsanitize=undefined -fsanitize=unreachable -fsanitize=bool -fsanitize=enum -fsanitize=leak -fsanitize-recover=address". This is difficult to do with ECM_ENABLE_SANITIZERS, and the full project requires around 8.6GB.

**Comment by neo1973** [comment]
> do you get the proper executable with debuginfo enabled once the compilation/linkage is over?

Yes, works with both GCC and Clang. The leak reports also have proper file+line information and I don't have issues when debugging with GDB.

Usually I call CMake with `-DECM_ENABLE_SANITIZERS='address;leak;undefined'`. This includes `bounds`, `unreachable`, `bool` and `enum` which are auto selected when enabling `undefined`. So only `fsanitize-recover=address` isn't enabled that way which I believe is the better default.

> the full kodi directory requires around 8.6GB

Are you referring to the build directory? This is really big, for my CMake config its only around 2.8GiB.

**Comment by repojohnray** [comment]
Note: This PR should be closed and replaced by https://github.com/lrusak/xbmc/commit/d5f6e35c2b6668f4eb76b9b37cc63d4b133f7e8e

**Pull request closed without merging**